### PR TITLE
once more unto the decomposition breach, dear friends!

### PR DESCRIPTION
**Pull request by Sacha0**
...or close up the wall with our obscure initialisms!

This pull request is take three on the recent decompositions rework; take two was #27159, and take one was #26997.

Specifically, this pull request deprecates `lufact[!]`, `schurfact[!]`, `lqfact[!]`, `qrfact[!]`, `ldltfact[!]`, and `svdfact[!]` to `lu[!]`, `schur[!]`, `lq[!]`, `qr[!]`, `ldlt[!]`, and `svd[!]` respectively, clobbering existing methods of the latter functions.

Additionally, this pull request (will eventually) deprecate `eig`/`eigfact[!]` to `eigen[!]`, `chol[!]`/`cholfact[!]` to `cholesky[!]`, `bkfact[!]` to `bunchkaufman[!]`, and `hessfact[!]` to `hessenberg[!]`.

To facilitate this deprecation, this pull request adds destructuring via iteration to some of the decomposition objects, and (immediately deprecated) `getindex(S::($DecompositionType), i::Integer)` methods that yield the decomposition components produced by iteration.

Pending items:
- [x] Deprecate `bkfact[!]` to `bunchkaufman[!]`
- [x] Deprecate `hessfact[!]` to `hessenberg[!]`
- [X] Deprecate `eigfact[!]` to `eigen[!]`
- [X] Deprecate `cholfact[!]` to `cholesky[!]`
- [X] Deprecate `eig` to `eigen`
- [ ] Deprecate `chol[!]` to `cholesky[!]`
- [X] Deprecate transitional `getindex` methods

Best!

### CODE REVIEW

**Comment by Sacha0** [comment]
To my surprise and delight, I was able to get everything done this evening but the gentler `chol[!]` -> `cholesky[!]` deprecation. I hope to eke out time for any issues CI catches tomorrow morning, and will try for the preceding last deprecation then as well. If I don't manage that last commit prior to other obligations, I will try to eke out additional time between those obligations, but otherwise pass the torch to `@mbauman` till Thursday :).

**Comment by fredrikekre** [comment]
Feel free to cherry pick from https://github.com/JuliaLang/julia/commits/breakmore

**Comment by Sacha0** [comment]
Incorporated Matt's `getindex` `@inline` removals into the relevant earlier commit, fixed the one CI failure in the `SparseArrays` suite, and removed the errant whitespace Matt found, so all commits up to "Expand pull request numbers for some news items" should be passing CI now.

Had a brief look at one of Fred's commits: I am not certain that `sqrt` is generally substitutable for `chol` in `choltype`, hence in the relevant commit of #27159 I kept `chol[!]` around internally as `legacychol[!]` for the moment. Might be overly conservative though :).

Out of time now, hope to find bits of time again in a couple of hours. Thanks all and best!

File: stdlib/LinearAlgebra/src/deprecated.jl [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# This file is a part of Julia. License is MIT: https://julialang.org/license
+ # This file is a part of Julia. License is MIT: https://julialang.org/license
```
[the remark this answers is not in the document]
Comment (Sacha0) [reply stdlib/LinearAlgebra/src/deprecated.jl]: Good catch! Should be fixed :).

File: stdlib/LinearAlgebra/src/deprecated.jl [hunk 2]
```diff
@@ -7,6 +7,7 @@ using Base: @deprecate, depwarn
 @deprecate cond(F::LinearAlgebra.LU, p::Integer) cond(convert(AbstractArray, F), p)
 
 # PR #22188
+export cholfact, cholfact!
```
[the remark this answers is not in the document]
Comment (Sacha0) [reply stdlib/LinearAlgebra/src/deprecated.jl]: Indeed so :). I had forgotten at first, and once I remembered halfway through I decided to proceed consistently. Please feel welcome to remove the explicit exports!

**Comment by simonbyrne** [comment]
It should be possible to deprecate `chol(X)` to `cholesky(X).U`

File: stdlib/LinearAlgebra/src/deprecated.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbauman) [reply stdlib/LinearAlgebra/src/deprecated.jl]: Not a major concern by any means — just thought I'd mention it as I was reviewing.

**Comment by fredrikekre** [comment]
> I am not certain that `sqrt` is generally substitutable for `chol` in `choltype`

Maybe not, but we should only have to keep the legacy_chol method for `Number`, since `choltype` calls `promote_type(..., Float32)` which only makes sense for `Number`s (right?). I think it boils down to redefining `choltype` as:
```
function choltype(A)
    x = one(eltype(A))
    promote_type(typeof(x), typeof(sqrt(abs(real(x)))), Float32)
end
```
and then we don't need `chol` or `legacychol`.

> It should be possible to deprecate `chol(X)` to `cholesky(X).U`

62e0b7e75c1008120856955c2665e90b80377475

**Comment by StefanKarpinski** [comment]
Failures are related to Pkg3 being renamed and/or timeouts.

**Comment by Jutho** [comment]
I very much like the new factorization interface and all the work that has been done here. Congrats to everbody involved. 

I do have one question. I tried to look for a discussion about the name change `eig` to `eigen`. Why was this decided or deemed necessary? `svd` is also a three letter word so that can be it. I like the name changes from `chol` to `cholesky` and the full name `hessenberg`, but `eig` seems so standard. It's also used by Matlab and Numpy. In fact, if it were not for `eigen`, the names would be identical to Numpy/Scipy: `numpy.linalg.eig`, `numpy.linalg.svd`, `numpy.linalg.cholesky`, `scipy.linalg.hessenberg`, `scipy.linalg.schur`.

**Comment by nalimilan** [comment]
See also https://discourse.julialang.org/t/eigen-the-odd-duck/14451.

**Comment by KristofferC** [comment]
There are still no docs for `iterate` on factorization objects?

**Pull request merged**